### ISSUE TITLE
[DATA] v4 Training Run #1: Echo Baseline (D=256, N=6, 650/10000 steps)

### ISSUE THREAD

**Issue by Kenessy** [VRAXION/VRAXION#107]
# v4 Training Run #1 — Echo Task Baseline

First sustained training run of INSTNCT v4. Ran on **RTX 4070 Ti SUPER 16GB**, echo task (byte prediction, embed_mode=true).

## Config at Time of Run

| Param | Value |
|---|---|
| D (embed_dim) | 256 |
| M (ring slots) | 256 |
| N (experts) | 6 |
| R (attention radius) | 2 |
| S (context scale) | 0.05 |
| kernel_mode | vshape |
| seq_len | 128 |
| batch_size | 16 |
| lr | 1e-3 |
| warmup_steps | 200 |
| max_grad_norm | 1.0 |
| embed_mode | true (byte, CE loss) |
| **Total params** | **528,134** |

## Results Summary

| Metric | Value |
|---|---|
| Steps completed | 630 of 10,000 |
| Wall time | 29.4 minutes (1762s) |
| Samples seen | 10,080 |
| Best loss | **2.108** (checkpoint tracker) |
| Best accuracy | **45.5%** at step 380 |
| Final loss | 2.365 at step 630 |
| Final accuracy | 38.5% at step 630 |

Random baseline for byte prediction = 1/256 = 0.39%. Reaching 45.5% means the model learned meaningful pattern prediction.

## Loss Curve (sampled)

```
Step  |  Loss  |  Acc%  |  LR       | Phase
------|--------|--------|-----------|------
    1 |  5.587 |   0.2% | 0.000005  | warmup start
   50 |  4.772 |  22.2% | 0.000250  | warmup
  100 |  3.000 |  30.5% | 0.000500  | warmup
  200 |  2.722 |  29.2% | 0.001000  | warmup end
  250 |  2.596 |  34.4% | 0.001000  | cosine decay
  340 |  2.332 |  40.8% | 0.001000  | improving
  380 |  2.147 |  45.5% | 0.000999  | *** PEAK ***
  400 |  2.338 |  37.7% | 0.000999  | oscillating
  540 |  2.259 |  41.7% | 0.000997  | second wind
  630 |  2.365 |  38.5% | 0.000995  | end
```

## Observations

1. **Rapid warmup phase (0-100):** Loss drops from 5.6 to 3.0, accuracy jumps from 0% to 30%. The model learns the basic echo pattern quickly.
2. **Peak at step 380:** Best loss 2.147, accuracy 45.5%. After this point, performance oscillates.
3. **Post-peak oscillation (380-630):** Loss bounces between 2.1-2.8, accuracy 32-45%. Suggests LR may be too high for fine-tuning, or batch_size=16 creates noisy gradients.
4. **Not VRAM-limited:** 528K params at D=256 uses only ~574MB. The bottleneck is the sequential T×N Python loop (128 timesteps × 6 experts per batch step).
5. **Run was stopped early** (630/10000) — convergence plateau was visible by step 400.

## Checkpoint Inventory

13 checkpoints saved (every 50 steps): `ckpt_step_000050.pt` through `ckpt_step_000650.pt` + `ckpt_latest.pt`. Total ~80MB.

Checkpoint format (CKPT_VERSION=1):
- `model_state`, `optimizer_state`, `config`, `step`, `best_loss`, `timestamp`

## What This Run Informed

- Issue #106 sweep data confirmed D=256 optimal at N=6, but N=2 nearly as good at 3x speed
- Sequential loop identified as primary bottleneck → precompute optimization (40% speedup)
- VRAM headroom discovery → hidden_dim/slot_dim split architecture (separate issue)

## Raw CSV

<details>
<summary>Full train_log.csv (click to expand)</summary>

```csv
step,raw_loss,masked_loss,accuracy,masked_acc,lr,elapsed_s,samples_seen,mask_frac
1,5.587114,5.587114,0.002441,0.002441,0.000005,2.8,16,1.0000
10,5.553921,5.553921,0.002441,0.002441,0.000050,29.1,160,1.0000
20,5.459164,5.459164,0.006348,0.006348,0.000100,59.3,320,1.0000
30,5.339748,5.339748,0.022949,0.022949,0.000150,90.2,480,1.0000
40,5.028507,5.026515,0.104492,0.104799,0.000200,115.8,640,0.9971
50,4.772491,4.769664,0.221680,0.222331,0.000250,150.6,800,0.9971
60,4.552114,4.552114,0.199219,0.199219,0.000300,178.5,960,1.0000
70,4.293698,4.293698,0.208496,0.208496,0.000350,204.1,1120,1.0000
80,3.755602,3.755602,0.237793,0.237793,0.000400,229.6,1280,1.0000
90,3.321660,3.321660,0.281738,0.281738,0.000450,257.7,1440,1.0000
100,3.000151,3.000151,0.305176,0.305176,0.000500,284.4,1600,1.0000
110,2.987558,2.987558,0.292969,0.292969,0.000550,310.8,1760,1.0000
120,2.940458,2.940458,0.295410,0.295410,0.000600,340.9,1920,1.0000
130,2.767403,2.767403,0.330078,0.330078,0.000650,367.3,2080,1.0000
140,2.828486,2.828486,0.333008,0.333008,0.000700,392.8,2240,1.0000
150,3.030948,3.030948,0.290527,0.290527,0.000750,419.3,2400,1.0000
160,3.076240,3.076240,0.266602,0.266602,0.000800,447.9,2560,1.0000
170,2.975780,2.975780,0.290039,0.290039,0.000850,477.6,2720,1.0000
180,2.883180,2.869018,0.283691,0.284525,0.000900,506.0,2880,0.9971
190,2.640687,2.640687,0.336914,0.336914,0.000950,534.7,3040,1.0000
200,2.721846,2.721846,0.291504,0.291504,0.001000,563.3,3200,1.0000
210,2.626334,2.611259,0.346191,0.347209,0.001000,592.5,3360,0.9971
220,2.599096,2.599096,0.341797,0.341797,0.001000,620.4,3520,1.0000
230,2.597327,2.597327,0.350098,0.350098,0.001000,648.2,3680,1.0000
240,2.693867,2.693867,0.334473,0.334473,0.001000,675.9,3840,1.0000
250,2.595516,2.595516,0.344238,0.344238,0.001000,704.1,4000,1.0000
260,2.631724,2.609914,0.346680,0.348039,0.001000,732.8,4160,0.9961
270,2.605175,2.605175,0.349121,0.349121,0.001000,761.7,4320,1.0000
280,2.579449,2.579449,0.349121,0.349121,0.001000,791.0,4480,1.0000
290,2.553102,2.553102,0.347168,0.347168,0.001000,820.7,4640,1.0000
300,2.768457,2.768457,0.312012,0.312012,0.001000,849.2,4800,1.0000
310,2.598475,2.598475,0.320801,0.320801,0.001000,877.5,4960,1.0000
320,2.497930,2.476120,0.358398,0.359452,0.001000,908.1,5120,0.9971
330,2.654978,2.654978,0.325684,0.325684,0.001000,937.6,5280,1.0000
340,2.331895,2.331895,0.408203,0.408203,0.001000,967.4,5440,1.0000
350,2.406238,2.406238,0.391113,0.391113,0.000999,995.1,5600,1.0000
360,2.575125,2.575125,0.321289,0.321289,0.000999,1022.5,5760,1.0000
370,2.487838,2.487838,0.354004,0.354004,0.000999,1050.0,5920,1.0000
380,2.147085,2.147085,0.454590,0.454590,0.000999,1077.0,6080,1.0000
390,2.651050,2.651050,0.327148,0.327148,0.000999,1106.0,6240,1.0000
400,2.337870,2.337870,0.377441,0.377441,0.000999,1134.4,6400,1.0000
410,2.779400,2.762949,0.323730,0.324682,0.000999,1163.8,6560,0.9971
420,2.484378,2.484378,0.391113,0.391113,0.000999,1192.0,6720,1.0000
430,2.508675,2.489122,0.361816,0.362880,0.000999,1221.0,6880,0.9971
440,2.442317,2.420776,0.377930,0.379040,0.000999,1249.6,7040,0.9971
450,2.485325,2.464680,0.346680,0.347698,0.000998,1277.8,7200,0.9971
460,2.532636,2.532636,0.342285,0.342285,0.000998,1306.9,7360,1.0000
470,2.388791,2.388791,0.372559,0.372559,0.000998,1335.6,7520,1.0000
480,2.378397,2.378397,0.388672,0.388672,0.000998,1364.5,7680,1.0000
490,2.495731,2.495731,0.360352,0.360352,0.000998,1393.0,7840,1.0000
500,2.369845,2.347513,0.374512,0.375612,0.000998,1420.6,8000,0.9971
510,2.394989,2.394989,0.387695,0.387695,0.000998,1447.1,8160,1.0000
520,2.619623,2.619623,0.366211,0.366211,0.000997,1475.6,8320,1.0000
530,2.367718,2.367718,0.411621,0.411621,0.000997,1500.8,8480,1.0000
540,2.259202,2.259202,0.416504,0.416504,0.000997,1529.2,8640,1.0000
550,2.530689,2.510778,0.370605,0.371694,0.000997,1556.4,8800,0.9971
560,2.315606,2.315606,0.402344,0.402344,0.000997,1582.4,8960,1.0000
570,2.257867,2.257867,0.412109,0.412109,0.000997,1608.2,9120,1.0000
580,2.389610,2.389610,0.378418,0.378418,0.000996,1635.5,9280,1.0000
590,2.480930,2.480930,0.360840,0.360840,0.000996,1661.2,9440,1.0000
600,2.342863,2.342863,0.392090,0.392090,0.000996,1686.6,9600,1.0000
610,2.328782,2.328782,0.388184,0.388184,0.000996,1712.2,9760,1.0000
620,2.381824,2.381824,0.398926,0.398926,0.000995,1737.2,9920,1.0000
630,2.364691,2.364691,0.385254,0.385254,0.000995,1761.5,10080,1.0000
```

</details>
